# Chapter 4 Python Practice Notebook

Angie Crews

In [2]:
import pandas as pd
from sqlalchemy import create_engine

print("pandas version:", pd.__version__)
print("Imports successful!")

pandas version: 3.0.5
Imports successful!


## Connect to PostgreSQL

Create a SQLAlchemy engine for connecting Python to the PostgreSQL database.

In [4]:
from getpass import getpass
from urllib.parse import quote_plus

password = getpass("Enter your PostgreSQL password: ")

engine = create_engine(
    f"postgresql+psycopg://postgres:{quote_plus(password)}@localhost:5432/sqlda"
)

print("Database engine created.")

Database engine created.


In [5]:
from sqlalchemy import text

with engine.connect() as connection:
    result = connection.execute(text("SELECT COUNT(*) FROM customers;"))
    customer_count = result.scalar()

print("Customer count:", customer_count)

Customer count: 50000


## Read SQL Data into pandas

Load a sample of customer records from PostgreSQL into a pandas DataFrame.

In [6]:
query = """
SELECT
    customer_id,
    first_name,
    last_name,
    city,
    state
FROM customers
LIMIT 10;
"""

customers_df = pd.read_sql(query, engine)

customers_df

,customer_id,first_name,last_name,city,state
0,716,Jarred,Bester,None,None
1,1228,Ag,Smerdon,None,None
2,1876,Giuditta,Eim,None,None
3,1991,Nichole,Rosle,None,None
4,2275,Chic,Bryning,None,None
5,2360,Jessee,Lytell,None,None
6,2552,Tova,Simao,None,None
7,3050,Huberto,Colerick,None,None
8,3236,Sherwynd,Lammert,None,None
9,3615,Hayyim,Tuftin,None,None


## Inspect the DataFrame

Review the DataFrame shape, column names, and data types.

In [7]:
print("Shape:", customers_df.shape)
print("\nColumn names:")
print(customers_df.columns)

print("\nData types:")
print(customers_df.dtypes)

Shape: (10, 5)

Column names:
Index(['customer_id', 'first_name', 'last_name', 'city', 'state'], dtype='str')

Data types:
customer_id     int64
first_name        str
last_name         str
city           object
state          object
dtype: object


## Check for Missing Values

Count missing values in each column of the sample DataFrame.

In [8]:
customers_df.isnull().sum()

customer_id     0
first_name      0
last_name       0
city           10
state          10
dtype: int64

## Check Missing Values in the Full Customer Table

Load the complete customer table and count missing values in each column.

In [9]:
query = """
SELECT
    customer_id,
    first_name,
    last_name,
    city,
    state
FROM customers;
"""

all_customers_df = pd.read_sql(query, engine)

print("Total customers:", len(all_customers_df))
print("\nMissing values:")
print(all_customers_df.isnull().sum())

Total customers: 50000

Missing values:
customer_id       0
first_name        0
last_name         0
city           5467
state          5467
dtype: int64


## Calculate the Percentage of Missing Values

Calculate the percentage of missing values in each column of the full customer table.

In [10]:
missing_percent = all_customers_df.isnull().sum() / len(all_customers_df) * 100

missing_percent

customer_id     0.000
first_name      0.000
last_name       0.000
city           10.934
state          10.934
dtype: float64

## Filter Customers with Location Data

Remove records with missing city or state values so the remaining data can be analyzed by location.

In [11]:
customers_with_location = all_customers_df.dropna(subset=["city", "state"])

print("Original customers:", len(all_customers_df))
print("Customers with location:", len(customers_with_location))
print("Removed:", len(all_customers_df) - len(customers_with_location))

customers_with_location.head()

Original customers: 50000
Customers with location: 44533
Removed: 5467


,customer_id,first_name,last_name,city,state
89,23169,Joyous,Blezard,El Paso,TX
118,27966,Martita,Perks,Mobile,AL
173,38559,Moll,Humbie,York,PA
185,41900,Deborah,Turner,Charleston,WV
202,44707,Drugi,Reace,Tacoma,WA


## Count Customers by State

Count customers by state and display the ten states or territories with the most customers.

In [12]:
state_counts = (
    customers_with_location["state"].value_counts().sort_values(ascending=False)
)

state_counts.head(10)

state
CA    5038
TX    4865
FL    3748
NY    2395
OH    1656
VA    1555
DC    1447
PA    1419
GA    1251
IL    1094
Name: count, dtype: int64

## Customer Counts by State

The cleaned `customers_with_location` DataFrame contains customer counts for each non-null state. The pandas code returns the ten states or territories with the most customers:

| Rank | State | Customers |
| ---: | :--- | ---: |
| 1 | CA | 5,038 |
| 2 | TX | 4,865 |
| 3 | FL | 3,748 |
| 4 | NY | 2,395 |
| 5 | OH | 1,656 |
| 6 | VA | 1,555 |
| 7 | DC | 1,447 |
| 8 | PA | 1,419 |
| 9 | GA | 1,251 |
| 10 | IL | 1,094 |

### Pandas

```python
state_counts = (
    customers_with_location["state"]
    .value_counts()
    .sort_values(ascending=False)
)

state_counts.head(10)
```

### Equivalent SQL

```sql
SELECT state, COUNT(*) AS customer_count
FROM customers
WHERE city IS NOT NULL
  AND state IS NOT NULL
GROUP BY state
ORDER BY customer_count DESC
LIMIT 10;
```

California has the largest number of customers in this result, followed by Texas and Florida.

## Filter Customers by State

Filter the cleaned customer data to keep customers from a specific state.

In [13]:
california_customers = customers_with_location[customers_with_location["state"] == "CA"]

print("California customers:", len(california_customers))

california_customers.head()

California customers: 5038


,customer_id,first_name,last_name,city,state
225,47687,Kiel,Celier,Visalia,CA
250,13,Gannon,Braker,San Diego,CA
252,15,Nichols,Espinay,Brea,CA
293,55,Pacorro,Lainge,Modesto,CA
303,65,Artair,Wholesworth,Sacramento,CA


## Compare SQL and pandas Filtering

Both SQL and pandas can filter data based on a condition.

### SQL

```sql
SELECT *
FROM customers
WHERE state = 'CA';
```

### pandas

```python
customers_with_location[
    customers_with_location["state"] == "CA"
]
```

In [14]:
customers_with_location[customers_with_location["state"] == "CA"]

,customer_id,first_name,last_name,city,state
225,47687,Kiel,Celier,Visalia,CA
250,13,Gannon,Braker,San Diego,CA
252,15,Nichols,Espinay,Brea,CA
293,55,Pacorro,Lainge,Modesto,CA
303,65,Artair,Wholesworth,Sacramento,CA
...,...,...,...,...,...
49956,49956,Keefe,Gooday,Van Nuys,CA
49962,49963,Tierney,Brake,San Francisco,CA
49979,49980,Orella,Eldredge,Bakersfield,CA
49986,49987,Maurita,Guyers,Fresno,CA


## Sort Customers

Use pandas `sort_values()` to sort customer data, similar to using `ORDER BY` in SQL.

### Equivalent SQL

```sql
SELECT *
FROM customers
WHERE city IS NOT NULL
  AND state IS NOT NULL
ORDER BY last_name
LIMIT 10;
```

In [15]:
sorted_customers = customers_with_location.sort_values(by="last_name")

sorted_customers.head(10)

,customer_id,first_name,last_name,city,state
22950,22761,Alair,A'field,Philadelphia,PA
26544,26384,Halette,A'field,Des Moines,IA
9477,9231,Dilly,Aaron,Crawfordsville,IN
28257,28110,Joscelin,Aarons,Colorado Springs,CO
38010,37941,Talbert,Aarons,Phoenix,AZ
37498,37424,Gerri,Aasaf,Washington,DC
13852,13719,Thia,Abad,Buffalo,NY
37267,37192,Prescott,Abad,Fort Lauderdale,FL
41746,41689,Sheila,Abadam,Raleigh,NC
43618,43575,Lay,Abadam,New Castle,PA


## Select Specific Columns

Select only the customer name, city, and state columns from the DataFrame. This is similar to selecting specific columns with `SELECT` in SQL.

### Equivalent SQL

```sql
SELECT first_name, last_name, city, state
FROM customers
WHERE city IS NOT NULL
  AND state IS NOT NULL
LIMIT 10;
```

In [16]:
selected_columns = customers_with_location[["first_name", "last_name", "city", "state"]]

selected_columns.head(10)

,first_name,last_name,city,state
89,Joyous,Blezard,El Paso,TX
118,Martita,Perks,Mobile,AL
173,Moll,Humbie,York,PA
185,Deborah,Turner,Charleston,WV
202,Drugi,Reace,Tacoma,WA
225,Kiel,Celier,Visalia,CA
228,Clarine,Miall,Reston,VA
237,Hermia,Darinton,Newark,DE
239,Ode,Stovin,Saint Louis,MO
240,Braden,Jordan,Pensacola,FL


## Filter Using Multiple Conditions

Filter the customer data using more than one condition. This is similar to using `AND` in a SQL `WHERE` clause.

### Equivalent SQL

```sql
SELECT *
FROM customers
WHERE state = 'TX'
  AND city = 'El Paso';
```

In [17]:
texas_el_paso = customers_with_location[
    (customers_with_location["state"] == "TX")
    & (customers_with_location["city"] == "El Paso")
]

print("Customers in El Paso, TX:", len(texas_el_paso))

texas_el_paso.head(10)

Customers in El Paso, TX: 713


,customer_id,first_name,last_name,city,state
89,23169,Joyous,Blezard,El Paso,TX
247,10,Barbara-anne,Gowlett,El Paso,TX
329,91,Tamqrah,O' Neligan,El Paso,TX
347,108,Gael,Hecks,El Paso,TX
371,132,Orrin,Evennett,El Paso,TX
385,145,Sandy,Brech,El Paso,TX
491,249,Orrin,Tocher,El Paso,TX
594,354,Phyllys,Kenforth,El Paso,TX
639,399,Orlan,Hasty,El Paso,TX
658,418,Benedetto,Pawley,El Paso,TX


## Group Customers by State

Use pandas `groupby()` to group customers by state and count the number of customers in each group. This is similar to using `GROUP BY` and `COUNT()` in SQL.

### Equivalent SQL

```sql
SELECT state, COUNT(*) AS customer_count
FROM customers
WHERE state IS NOT NULL
GROUP BY state
ORDER BY customer_count DESC
LIMIT 10;
```

In [18]:
customers_by_state = (
    customers_with_location.groupby("state")
    .size()
    .reset_index(name="customer_count")
    .sort_values("customer_count", ascending=False)
)

customers_by_state.head(10)

,state,customer_count
4,CA,5038
43,TX,4865
9,FL,3748
34,NY,2395
35,OH,1656
45,VA,1555
7,DC,1447
38,PA,1419
10,GA,1251
14,IL,1094


## Customer Data Summary

Summarize the customer data after removing records with missing location information.

In [19]:
print("Total customers:", len(all_customers_df))
print("Customers with location:", len(customers_with_location))
print("States represented:", customers_with_location["state"].nunique())
print("Largest customer state:", customers_by_state.iloc[0]["state"])
print("Customers in largest state:", customers_by_state.iloc[0]["customer_count"])

Total customers: 50000
Customers with location: 44533
States represented: 51
Largest customer state: CA
Customers in largest state: 5038


## Conclusion

This practice notebook connected Python to PostgreSQL and used pandas to analyze customer data. The workflow included loading SQL data into a DataFrame, checking the data, finding missing values, filtering and sorting records, selecting columns, and grouping data.

The customer table contained 50,000 records. Of these, 5,467 customers were missing city and state information, leaving 44,533 customers with complete location data. California had the largest number of customers, with 5,038.

This practice demonstrated how SQL and pandas work together: SQL retrieves data from the database, while pandas cleans, organizes, and analyzes the results in Python.